client/my-sites/checkout/cart/test/cart-coupon.js

```diff
@@ -14,10 +14,11 @@ import React from 'react';
  * Internal dependencies
  */
 import { CartCoupon } from '../cart-coupon';
-import { applyCoupon } from 'lib/upgrades/actions';
+import { applyCoupon, removeCoupon } from 'lib/upgrades/actions';
 
 jest.mock( 'lib/upgrades/actions', () => ( {
 	applyCoupon: jest.fn( () => {} ),
+	removeCoupon: jest.fn( () => {} ),
 } ) );
 
 jest.mock( 'lib/analytics', () => ( {
```

Comment:
Just a note as I know this wasn't added in this PR. If we avoid hitting `lib/analytics` directly and instead dispatch analytics actions from components as was [already suggested](https://github.com/Automattic/wp-calypso/pull/25433/files#r194632132), we can avoid mocking this lib, which would be a _good thing_.

Reply:
Thank you. I've added a `beforeEach()` as you suggested and I took some of your other ideas and implemented them in the tests alongside Redux actions for `applyCoupon` and `removeCoupon` suggested by others. I also used `withAnalytics()`.

However, I'm not seeing a way to get rid of these mocks because we need them to confirm the tests succeed; i.e., to check if the underlying actions were in fact triggered by test events. Given recent revisions, do you see a way to remove these mocks?

Reply:
The `tracks` mock shouldn't be necessary now, as tracks events are going through redux actions 👍 

---

> do you see a way to remove these mocks?

We need to spy on `applyCoupon` and `removeCoupon`, not change their functionality, is that right?

You can just pass a spy when necessary:

<details><summary>patch</summary>

```patch
diff --git a/client/my-sites/checkout/cart/test/cart-coupon.js b/client/my-sites/checkout/cart/test/cart-coupon.js
index 2f0aeb5eca..4a95b39c19 100644
--- a/client/my-sites/checkout/cart/test/cart-coupon.js
+++ b/client/my-sites/checkout/cart/test/cart-coupon.js
@@ -7,7 +7,7 @@
  * External dependencies
  */
 import { mount, shallow } from 'enzyme';
-import { identity } from 'lodash';
+import { identity, noop } from 'lodash';
 import React from 'react';
 
 /**
@@ -31,8 +31,8 @@ jest.mock( 'lib/analytics', () => ( {
 } ) );
 
 const props = {
-	applyCoupon,
-	removeCoupon,
+	applyCoupon: noop,
+	removeCoupon: noop,
 	translate: identity,
 };
 
@@ -47,11 +47,6 @@ describe( 'cart-coupon', () => {
 		coupon: '',
 	};
 
-	beforeEach( function() {
-		applyCoupon.mockReset();
-		removeCoupon.mockReset();
-	} );
-
 	describe( 'General behavior', () => {
 		test( 'Should not blow up', () => {
 			const component = shallow( <CartCoupon { ...props } cart={ cart } /> );
@@ -107,9 +102,11 @@ describe( 'cart-coupon', () => {
 		} );
 
 		test( 'Should apply a coupon when form is submitted', () => {
+			const applyCouponSpy = jest.fn();
 			const component = shallow(
 				<CartCoupon
 					{ ...props }
+					applyCoupon={ applyCouponSpy }
 					cart={ {
 						...cart,
 						is_coupon_applied: false,
@@ -122,8 +119,8 @@ describe( 'cart-coupon', () => {
 				couponInputValue: 'CODE15',
 			} );
 			component.find( '.cart__form' ).simulate( 'submit', event );
-			expect( applyCoupon.mock.calls.length ).toBe( 1 );
-			expect( applyCoupon.mock.calls[ 0 ][ 0 ] ).toBe( 'CODE15' );
+			expect( applyCouponSpy ).toHaveBeenCalledTimes( 1 );
+			expect( applyCouponSpy ).toHaveBeenCalledWith( 'CODE15' );
 		} );
 
 		test( 'Should disallow submission when form is currently being submitted', () => {
```

</details>

---

Something like the following (untested) should get off `lib/analytics` for this component and allow the entire `lib/analytics` mock to be removed:

<details><summary>patch</summary>

```patch
diff --git a/client/my-sites/checkout/cart/cart-coupon.jsx b/client/my-sites/checkout/cart/cart-coupon.jsx
index 9ee8b82e0c..0124ac64cb 100644
--- a/client/my-sites/checkout/cart/cart-coupon.jsx
+++ b/client/my-sites/checkout/cart/cart-coupon.jsx
@@ -13,8 +13,7 @@ import { localize } from 'i18n-calypso';
  * Internal dependencies
  */
 import Button from 'components/button';
-import analytics from 'lib/analytics';
-import { withAnalytics, recordTracksEvent } from 'state/analytics/actions';
+import { withAnalytics, recordTracksEvent, recordGoogleEvent } from 'state/analytics/actions';
 import { applyCoupon, removeCoupon } from 'lib/upgrades/actions';
 
 export class CartCoupon extends React.Component {
@@ -117,9 +116,9 @@ export class CartCoupon extends React.Component {
 		this.setState( { isCouponFormShowing: ! this.state.isCouponFormShowing } );
 
 		if ( this.state.isCouponFormShowing ) {
-			analytics.ga.recordEvent( 'Upgrades', 'Clicked Hide Coupon Code Link' );
+			recordGoogleEvent( 'Upgrades', 'Clicked Hide Coupon Code Link' );
 		} else {
-			analytics.ga.recordEvent( 'Upgrades', 'Clicked Show Coupon Code Link' );
+			recordGoogleEvent( 'Upgrades', 'Clicked Show Coupon Code Link' );
 		}
 	};
 
@@ -166,6 +165,7 @@ export class CartCoupon extends React.Component {
 }
 
 const mapDispatchToProps = dispatch => ( {
+	recordGoogleEvent,
 	applyCoupon: coupon =>
 		dispatch(
 			withAnalytics(
diff --git a/client/my-sites/checkout/cart/test/cart-coupon.js b/client/my-sites/checkout/cart/test/cart-coupon.js
index 2f0aeb5eca..ff791821bd 100644
--- a/client/my-sites/checkout/cart/test/cart-coupon.js
+++ b/client/my-sites/checkout/cart/test/cart-coupon.js
@@ -21,15 +21,6 @@ jest.mock( 'lib/upgrades/actions', () => ( {
 	removeCoupon: jest.fn( () => {} ),
 } ) );
 
-jest.mock( 'lib/analytics', () => ( {
-	ga: {
-		recordEvent: () => {},
-	},
-	tracks: {
-		recordEvent: () => {},
-	},
-} ) );
-
 const props = {
 	applyCoupon,
 	removeCoupon,
diff --git a/npm-shrinkwrap.json b/npm-shrinkwrap.json
index b8d5886be0..e2ba35f56b 100644
Binary files a/npm-shrinkwrap.json and b/npm-shrinkwrap.json differ

```
</details>

This are nice janitorial changes to make, but it's probably best not to bloat this PR.

Reply:
Awesome! I was AFK for a few days last week, but I've now used those helpful examples you gave me. I was able to get rid of all the global mocks. Thank you again 👍